.github/workflows/CICD.yaml

```diff
@@ -8,53 +8,263 @@ on:
+          source ${{ github.workspace }}/.github-venv/bin/activate
+          ./node_modules/bin/pyright
+
+  Pytest:
+    name: Automated testing
+    runs-on: ubuntu-latest
+    needs:
+      - Checkout
+
+    env:
+      PYTHON_VERSION: "3.10"
+
+    if: ${{( success() && !cancelled() ) }}
 
-      - name: Test with pytest
+    steps:
+      - name: Set up Python ${{ env.PYTHON_VERSION }}
+        uses: actions/setup-python@v5
+        id: install_python
+        with:
+          python-version: ${{ env.PYTHON_VERSION }}
+
+      - uses: actions/cache/restore@v4
+        name: Restore run cache
+        id: restore-run-cache
+        with:
+          key: ${{ needs.Checkout.outputs.cache-key-run }}
+          path: ${{ github.workspace }}
+          fail-on-cache-miss: true
+
+      - name: Test with pytest and generate reports
         run: |
           echo Running pytest
-          # infra is not set up to support integration tests, so ignore postgresql
-          # etl has been updated, while the tests have not been
-          pytest --junitxml pytest.xml \
-                 -k "not acceptance"
+          source ${{ github.workspace }}/.github-venv/bin/activate
+          pytest -k "not acceptance"
+
+          coverage html
 
       - name: Output pytest report
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
+        if: ${{ always() }}
         with:
-          name: pytest-report
-          path: pytest.xml
+          name: pytest-and-coverage-report
+          path: |
+            pytest.xml
+            cov.xml
+            .coverage
+            htmlcov/
           retention-days: 1
           if-no-files-found: error
 
+  Style:
+    name: Style and formatting
+    runs-on: ubuntu-latest
+    needs:
+      - Checkout
+
+    env:
+      PYTHON_VERSION: "3.10"
+
+    if: ${{( success() && !cancelled() ) }}
+
+    steps:
+      - name: Set up Python ${{ env.PYTHON_VERSION }}
+        uses: actions/setup-python@v5
+        id: install_python
+        with:
+          python-version: ${{ env.PYTHON_VERSION }}
+
+      - uses: actions/cache/restore@v4
+        name: Restore run cache
+        id: restore-run-cache
+        with:
+          key: ${{ needs.Checkout.outputs.cache-key-run }}
+          path: ${{ github.workspace }}
+          fail-on-cache-miss: true
+
       - name: Check code style
         run: |
+          source .github-venv/bin/activate
           black --check src
 
       - name: Check import order
         run: |
+          source .github-venv/bin/activate
           isort --check-only src
+
+  Final-status-check:
+    name: Check workflow success
+    runs-on: ubuntu-latest
+    needs:
+      - Checkout
+      - Pyright
+      - Pytest
+      - Style
+    # this job should run regardless of success, failure, or skips,
+    # but not if the workflow is canceled. `always()` ignores canceled,
+    # and so we check for the canceled state.
+    if: ${{( always() && !cancelled() ) }}
```

Comment:
It kills me that both "canceled" and "cancelled" are acceptable, but it will kill me more if we use both.

Reply:
:D Good catch, thanks.